tf-IDEC/autoencoder.py

```diff
@@ -0,0 +1,95 @@
+"""Define and pretrain an autoencoder as defined in the DEC/IDEC papers."""
+
+import os
+import sys
+import logging
+import datetime
+
+import numpy as np
+import tensorflow as tf
+from tensorflow.examples.tutorials.mnist import input_data as tf_datasets
+
+from constants import (k, Nm, NUM_CHANNELS, IMAGE_SHAPE,
+                       PRETRAIN_SAVE_EVERY_N_EPOCHS, PRETRAIN_ITERATIONS, AE_LOGDIR)
+from utils import load_mnist, add_train_ops, set_logging_verbosity
+
+
+def pretrain_autoencoder(placeholders):
+    """Pretrain the autoencoder end-to-end on MSE objective."""
+    inp, encode_fn, decode_fn = build_autoencoder_graph(placeholders)
+
+    saver = tf.train.Saver()
+    sess = tf.Session()
+
+    # AE loss is just MSE between (flattened) input and decoder output
+    with tf.name_scope('loss'):
+        loss = tf.losses.mean_squared_error(inp, decode_fn)
+
+    # Collect trainable variables, define train op as Adam and init the session
+    trainable_var = tf.trainable_variables()
+    train_op = add_train_ops(loss, trainable_var)
+
+    # Init graph after adding train/loss ops
+    sess.run(tf.global_variables_initializer())
+
+    X, _ = load_mnist()
+
+    num_batches = len(X) // Nm
+    for epoch in range(PRETRAIN_ITERATIONS):
+
+        # Shuffle X before each epoch
+        np.random.shuffle(X)
+
+        # Train over all full batches; we shuffle so we don't have to worry
+        # about when a partial batch is left over
+        train_loss = 0
+        for batch_ix in range(num_batches):
+            # Get next batch for training, then reshape to be 4D
+            # (our first layer after the input layer will flatten the input)
+            batch_X = X[batch_ix * Nm: (batch_ix + 1) * Nm]
+            batch_X = np.reshape(batch_X, (Nm, NUM_CHANNELS, *IMAGE_SHAPE))
```

Comment:
I might have missed something but X is already reshaped in `load_mnist()`

Reply:
Good call, removed and updated comment